src/compiler/crystal/codegen/primitives.cr

```diff
@@ -1299,41 +1296,18 @@ class Crystal::CodeGenVisitor
     end
   end
 
-  {% if LibLLVM::IS_LT_130 %}
-    def validate_atomic_cmpxchg_ordering(success_node, success_ordering, failure_node, failure_ordering)
-      success_node.raise "must be atomic" if success_ordering.not_atomic?
-      failure_node.raise "must be atomic" if failure_ordering.not_atomic?
+  def validate_atomic_cmpxchg_ordering(success_node, success_ordering, failure_node, failure_ordering)
+    success_node.raise "must be atomic" if success_ordering.not_atomic?
+    success_node.raise "cannot be unordered" if success_ordering.unordered?
 
-      success_node.raise "cannot be unordered" if success_ordering.unordered?
-      failure_node.raise "cannot be unordered" if failure_ordering.unordered?
+    failure_node.raise "must be atomic" if failure_ordering.not_atomic?
+    failure_node.raise "cannot be unordered" if failure_ordering.unordered?
+    failure_node.raise "cannot include release semantics" if failure_ordering.release? || failure_ordering.acquire_release?
 
-      case success_ordering
-      when .monotonic?
-        unless failure_ordering.monotonic?
-          failure_node.raise "shall be no stronger than success ordering"
-        end
-      when .acquire?, .release?
-        case failure_ordering
-        when .release?
-          failure_node.raise "cannot include release semantics"
-        when .acquire_release?, .sequentially_consistent?
-          failure_node.raise "shall be no stronger than success ordering"
-        end
-      when .acquire_release?
-        case failure_ordering
-        when .release?, .acquire_release?
-          failure_node.raise "cannot include release semantics"
-        when .sequentially_consistent?
-          failure_node.raise "shall be no stronger than success ordering"
-        end
-      when .sequentially_consistent?
-        case failure_ordering
-        when .release?, .acquire_release?
-          failure_node.raise "cannot include release semantics"
-        end
-      end
-    end
-  {% end %}
+    {% if LibLLVM::IS_LT_130 %}
+      failure_node.raise "shall be no stronger than success ordering" if failure_ordering > success_ordering
+    {% end %}
```

Comment:
**suggestion:** I'm wondering if we shouldn't activate this check on all LLVM versions for consistency.
If we want to keep it only for older versions, there should be a comment that explains why.

Reply:
_Thinking out loud_: they're two sides of the same coin, but don't have the same goal? The compiler side is making sure the codegen is correct as per LLVM, which for atomics depends on the LLVM version, while the stdlib side needs to be consistent as per Crystal, regardless of the LLVM version?